File: 121. Best time to buy and sell stock/solution.py
# python solution in O(n) time
class Solution:
    '''
        Checking out the -> operator in python. It basically marks the return function convention.
        https://stackoverflow.com/questions/14379753/what-does-mean-in-python-function-definitions
    '''
    def maxProfit(self, prices: List[int]) -> int:

        # check if the prices array is not empty
        if not prices:
            return 0
        # we generally use the snake case in python
        minimum_price, maximum_profit =  price[0], 0
        for i in range(1, len(prices)):
            minimum_price =  min(minimum_price, prices[i])
            maximum_profit =  max(maximum_profit, prices[i]-minimum_price)
        return mx

File: 121. Best time to buy and sell stock/solution.cpp
// Time: O(n)
// Space: O(1)

#include <vector>
#include <algorithm>
using std::vector;
using std::max;
using std::min;


class Solution {
	public:
		int maxProfit(vector<int> &prices) {
			if (prices.empty()) return 0;

			/** https://stackoverflow.com/questions/23120856/use-of-curly-brackets-braces-around-a-variable-c*/
			int ret{0}, low{prices[0]};
			for (auto price: prices) {
				low = min(low,price);
				ret = max(ret,price-low);
			}
			return ret;
		}
};
/**
 * numeric_limits : https://en.cppreference.com/w/cpp/types/numeric_limits
 * range based iteration : *using the auto keyword* : https://www.geeksforgeeks.org/range-based-loop-c/
 */


